core/src/main/java/com/linecorp/armeria/server/throttling/BlockingTaskLimitingThrottlingStrategy.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.throttling;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.lang.reflect.Field;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.BlockingTaskExecutor;
+import com.linecorp.armeria.common.util.SettableIntSupplier;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * A {@link ThrottlingStrategy} that
+ * provides a throttling strategy based on the queue size of the {@link BlockingTaskExecutor}.
+ *
+ * <p>Note that the size of the queue and
+ * the limit of the {@link SettableIntSupplier} are not guaranteed to be atomic.
+ *
+ * @see ThrottlingStrategy#blockingTaskLimiting(int, String)
+ * @see ThrottlingStrategy#blockingTaskLimiting(SettableIntSupplier, String)
+ */
+final class BlockingTaskLimitingThrottlingStrategy<T extends Request> extends ThrottlingStrategy<T> {
+    private static final Logger logger = LoggerFactory.getLogger(BlockingTaskLimitingThrottlingStrategy.class);
+
+    private final SettableIntSupplier settableLimitSupplier;
+
+    BlockingTaskLimitingThrottlingStrategy(SettableIntSupplier settableLimitSupplier,
+                                           @Nullable String name) {
+        super(name);
+        checkArgument(settableLimitSupplier.getAsInt() > 0, "limit must be larger than zero");
+        this.settableLimitSupplier = settableLimitSupplier;
+    }
+
+    @Override
+    public CompletionStage<Boolean> accept(ServiceRequestContext ctx, T request) {
+        final ExecutorService executorService = ctx.blockingTaskExecutor();
+        final ThreadPoolExecutor executor =
+                unwrapThreadPoolExecutor(executorService,
+                                         executorService.getClass().getSuperclass().getSuperclass());
```

Comment:
- These reflections could be expensive to run every time `accept()` is called.
  Could we cache the pair of `ExecutorService` and `ThreadPoolExecutor`.
- Should we check whether the executor is `ThreadPoolExecutor` or `ForkJoinPool`?
For example:
```java
private final ConcurrentHashMap<ExecutorService, ThreadPoolExecutor> executorCache = ...;

public CompletionStage<Boolean> accept(ServiceRequestContext ctx, T request) {

    // A `ContextAwareExecutorService` is newly created if a new request is received.
    // We should unwrap the ContextAwareScheduledExecutorService to cache the underlying ExecutorService.
    final ExecutorService executorService = ctx.blockingTaskExecutor().withoutContext();
    if (executorService instanceof ThreadPoolExecutor) {
         ...
    } else if (executorService instanceof ForkJoinPool) {
         ...
    } else {
        executorCache.computeIfAbsent(executorService, key -> unwrapThreadPoolExecutor(...));
        ...
    }
```

Reply:
1. I will cache it when the object is contruct
2. I'm sorry, but is there any reason to check `ForkJoinPool`? I thought that `ctx.blockingTaskExecutor()` is always `ContextAwareScheduledExecutorService`.

Reply:
> I'm sorry, but is there any reason to check ForkJoinPool? I thought that ctx.blockingTaskExecutor() is always ContextAwareScheduledExecutorService.

`ContextAwareScheduledExecutorService` wraps an actual working executor. Therefore, we should wrap the underlying executor to collect the metric of the executor.
The code example I suggested was wrong. It should call `unwrap()` after `withoutContext()` for a `BlockingTaskExecutor`
```diff
- final ExecutorService executorService = ctx.blockingTaskExecutor().withoutContext();
+ ExecutorService executorService = ctx.blockingTaskExecutor().withoutContext();
+ if (executorService instanceof BlockingTaskExecutor) {
+     executorService = ((BlockingTaskExecutor) executorService).unwrap();
+ }
```